lib/Conversion/TorchToTosa/TorchToTosaCustom.cpp

```diff
@@ -0,0 +1,143 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// Also available under a BSD-style license. See LICENSE.
+//
+//===----------------------------------------------------------------------===//
+
+#include "torch-mlir/Conversion/TorchToTosa/TorchToTosa.h"
+#include "torch-mlir/Conversion/TorchToTosa/TosaLegalizeCommon.h"
+#include "torch-mlir/Conversion/TorchToTosa/TosaLegalizeUtils.h"
+
+#include "../PassDetail.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "mlir/Dialect/Traits.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "torch-mlir/Dialect/Torch/IR/TorchOps.h"
+#include "torch-mlir/Dialect/Torch/Utils/Utils.h"
+#include "torch-mlir/Dialect/TorchConversion/IR/TorchConversionDialect.h"
+#include "torch-mlir/Dialect/TorchConversion/Transforms/BackendTypeConversion.h"
+
+using namespace mlir;
+using namespace mlir::torch;
+using namespace mlir::torch::Torch;
+
+namespace {
+
+// This defines a template to construct ops whose legalizations are
+// specialized.
+template <typename AtenOpT>
+class ConvertAtenOp : public OpConversionPattern<AtenOpT> {
+public:
+  using OpConversionPattern<AtenOpT>::OpConversionPattern;
+  using OpAdaptor = typename AtenOpT::Adaptor;
+  LogicalResult
+  matchAndRewrite(AtenOpT op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override;
+};
+
+template <>
+LogicalResult ConvertAtenOp<AtenSoftmaxIntOp>::matchAndRewrite(
```

Comment:
This is not a systematic conversion to `tosa::CustomOp`. By systematic, I mean that the pattern should apply to every op in the IR and it should convert to `tosa::CustomOp` without using any op-specific knowledge. In other words, the transformations should be done on `Operation *` objects.

Reply:
I thought we can do the #define INSERT_ATENOP_PATTERN(AtenOp) ...     kind of thing for every op just like what the TorchToTosa.cpp has done.

Reply:
No, we don't want to write a new pattern for every possible PyTorch op (this is what I meant in my previous comment when I mentioned that we don't want "per-op custom patterns for converting to tosa::CustomOp"). This type of transformation should be achievable with a single pattern that applies to all ops.

Reply:
> And Since the !torch.none kind of type conversion is not handled yet, and 'tosa.custom' op operand must be a tensor of number values.

If this PR does not yet plan on handling `!torch.none`, then you should choose a simpler op that does not take `!torch.none`s as inputs for the first PR.

> I don't think there is a general way to handle all ops.

If there is no general way of converting a good subset of PyTorch ops to `tosa::CustomOp`, then we should improve the interface to `tosa::CustomOp` first, otherwise this will bring little value to users. However, I do think you should already be able to handle a good subset of ops, namely, all ops that at compile time take either tensors as input or constants. I would focus on this and reject all other ops for now.

Reply:
Any suggestion how to handle the **!torch.none**?